### PR TITLE
create figures

### PR DESCRIPTION
**Pull request by ConfusedSatlan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,18 +18,18 @@ In the `main()` method we need to create an array of figures (the size of array
 
 For this purpose create two more classes:
 - `ColorSupplier` with `public String getRandomColor()` method - for generating random color, 
-- and `FigureSupplier` with `public Figure getRandomFigure()` method - for generating figure with random properties.
```
Comment (andrii-hoienko) [inline README.md:21]: Please do not edit and push README.md file

File: checklist.md [hunk 2]
```diff
@@ -67,7 +67,7 @@ public class FigureSupplier {
     public static final int FIGURE_COUNT = 5;
     private Random random = new Random();
 
-    public Figure getRandomFigure() {
```
Comment (andrii-hoienko) [inline checklist.md:70]: Please do not edit and push checklist.md file

File: src/main/java/core/basesyntax/Main.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package core.basesyntax;
+
+import core.basesyntax.figure.Figure;
+import core.basesyntax.tools.FigureSupplier;
+import java.util.Random;
+
+public class Main {
+    public static void main(String[] args) {
+
+        FigureSupplier figureCreater = new FigureSupplier();
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/Main.java:10]: This should be class level variable (https://mate-academy.github.io/jv-program-common-mistakes/java-core/abstract-class-interface/oop-advanced#think-about-which-variables-should-be-local-in-the-method-and-which-should-be-class-level)

File: src/main/java/core/basesyntax/Main.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import core.basesyntax.figure.Figure;
+import core.basesyntax.tools.FigureSupplier;
+import java.util.Random;
+
+public class Main {
+    public static void main(String[] args) {
+
+        FigureSupplier figureCreater = new FigureSupplier();
+        int sizeLists = new Random().nextInt(FigureSupplier.MIN_COUNT_FIGURE + 1)
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/Main.java:11]: Number of iterations over loop is constant and dont need to be calculated

File: src/main/java/core/basesyntax/behaviour/FigureDraw.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package core.basesyntax.behaviour;
+
+public interface FigureDraw {
+
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/behaviour/FigureDraw.java:4]: redundant empty line

File: src/main/java/core/basesyntax/figure/Figure.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
+package core.basesyntax.figure;
+
+import core.basesyntax.behaviour.FigureCreation;
+import core.basesyntax.behaviour.FigureDraw;
+import core.basesyntax.tools.Color;
+import java.util.Random;
+
+public abstract class Figure implements FigureDraw, FigureCreation {
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/figure/Figure.java:8]: Don’t use abstract classes to set behaviour for classes (https://mate-academy.github.io/jv-program-common-mistakes/java-core/abstract-class-interface/oop-advanced#dont-use-abstract-classes-to-set-behaviour-for-classes)
There is not need to store area in Figure

File: src/main/java/core/basesyntax/figure/Figure.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    protected static int randomSideFigure() {
+        return new Random().nextInt(MAX_SIDE_FIGURE) + 1;
+    }
+
+    public double getArea() {
+        return area;
+    }
+
+    public void setArea(double area) {
+
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/figure/Figure.java:23]: Pay attention to empty lines

File: src/main/java/core/basesyntax/figure/Figure.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getColor() {
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/figure/Figure.java:35]: This method should return color, in our program color is enum, so please return propriate value here

File: src/main/java/core/basesyntax/tools/ColorSupplier.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.tools;
+
+import java.util.Random;
+
+public class ColorSupplier {
+    public String getRandomColor() {
+        int colorRandomIndex = new Random().nextInt(Color.values().length);
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/tools/ColorSupplier.java:7]: Overloaded expression, please introduce some variables. Also Random should be class level variable

File: src/main/java/core/basesyntax/tools/FigureSupplier.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+                return new RightTriangle("rightTriangle");
+            case 4:
+                return new IsoscelesTrapezoid("isoscelesTrapezoid");
+            default:
+                return getDefaultFigure();
+        }
+    }
+
+    public Figure getDefaultFigure() {
+        return new Circle("circle",10, Color.WHITE);
```
Comment (andrii-hoienko) [inline src/main/java/core/basesyntax/tools/FigureSupplier.java:36]: All magic numbers in your code should be constants (https://mate-academy.github.io/jv-program-common-mistakes/java-core/abstract-class-interface/oop-advanced#all-magic-numbers-in-your-code-should-be-constants)

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Please read common mistakes more cerafully

File: src/main/java/core/basesyntax/figure/Figure.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public double getArea() {
+        return 0;
+    }
+
+    @Override
+    public void resultArea() {
+    }
```
Comment (Vasul-Pankiv) [inline src/main/java/core/basesyntax/figure/Figure.java:51]: do not implement interface methods in abstract class and pay more attention to comment (you already have the same comment) or ask a questions if you don't understand. We will be happy to help
```suggestion
```

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/figure/Figure.java (hunk 11 above)
Comment (ConfusedSatlan) [reply src/main/java/core/basesyntax/figure/Figure.java:51]: where should I put them?

File: src/main/java/core/basesyntax/behaviour/FigureDraw.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package core.basesyntax.behaviour;
+
+public interface FigureDraw {
```
[the remark this answers is not in the document]
Comment (ConfusedSatlan) [reply src/main/java/core/basesyntax/behaviour/FigureDraw.java:3]: FigureDrawer nWhat are the rules for choosing a name? I choose the naming of the execution of the variable, my name is drawing a figure in present, figure draw,  whats problem with that?ot

File: src/main/java/core/basesyntax/Main.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+import core.basesyntax.figure.Figure;
+import core.basesyntax.tools.FigureSupplier;
+
+public class Main {
+    private static FigureSupplier figureCreater = new FigureSupplier();
+
+    public static void main(String[] args) {
+        final int sizeLists = 6;
+        Figure[] figureLists = new Figure[sizeLists];
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Main.java:11]:
```suggestion
        final int arraySize = 6;
        Figure[] figures = new Figure[arraySize];
```
please use proper naming

File: src/main/java/core/basesyntax/behaviour/FigureAreaCalculator.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.behaviour;
+
+public interface FigureAreaCalculator {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/behaviour/FigureAreaCalculator.java:4]:
```suggestion
```
redundant empty line

File: src/main/java/core/basesyntax/behaviour/FigureCreator.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.behaviour;
+
+import core.basesyntax.figure.Figure;
+
+public interface FigureCreator {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/behaviour/FigureCreator.java:6]: same

File: src/main/java/core/basesyntax/behaviour/FigureAreaCalculator.java [hunk 16]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.behaviour;
+
+public interface FigureAreaCalculator {
+
+    double getArea();
+
+    void setArea(double area);
+
+    void resultArea();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/behaviour/FigureAreaCalculator.java:9]: let's remove these methods - you will see that we don't need them

File: src/main/java/core/basesyntax/behaviour/FigureAreaCalculator.java [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package core.basesyntax.behaviour;
+
+public interface FigureAreaCalculator {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/behaviour/FigureAreaCalculator.java:3]:
```suggestion
public interface AreaCalculator {
```
this interface provides method to get area for some object
in future, we could use it not only for figures

File: src/main/java/core/basesyntax/figure/Circle.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
+package core.basesyntax.figure;
+
+import core.basesyntax.tools.Color;
+
+public class Circle extends Figure {
+    private int radius;
+    private double area;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Circle.java:7]:
```suggestion
```
let's remove this field

the problem is, when we store the value of area - if we change radius, we will have to recalculate area as well
this takes additional effort that we can avoid

File: src/main/java/core/basesyntax/figure/Circle.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+import core.basesyntax.tools.Color;
+
+public class Circle extends Figure {
+    private int radius;
+    private double area;
+
+    public Circle(String name, Color color, int radius) {
+        super(name, color);
+        this.radius = radius;
+        resultArea();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Circle.java:12]:
```suggestion
```

File: src/main/java/core/basesyntax/figure/Circle.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+    public void setRadius(int radius) {
+        this.radius = radius;
+        resultArea();
+    }
+
+    @Override
+    public void resultArea() {
+        setArea(2 * Math.PI * radius * radius);
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Circle.java:43]:
```suggestion
    @Override
    public double getArea() {
        return 2 * Math.PI * radius * radius;
    }

    public int getRadius() {
        return radius;
    }

    public void setRadius(int radius) {
        this.radius = radius;
    }
```

File: src/main/java/core/basesyntax/figure/Circle.java [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public void draw() {
+        super.draw();
+        System.out.println(" area: "
+                + getArea() + " sq.units"
+                + ", radius: " + getRadius() + " units, "
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Circle.java:50]:
```suggestion
                + ", radius: " + radius + " units, "
```
no need to use getter when we can address field directly

File: src/main/java/core/basesyntax/figure/Figure.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+
+import core.basesyntax.behaviour.Drawable;
+import core.basesyntax.behaviour.FigureAreaCalculator;
+import core.basesyntax.tools.Color;
+
+public abstract class Figure implements Drawable, FigureAreaCalculator {
+    private Color colorFigure;
+    private String name;
+
+    Figure(String name, Color color) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Figure.java:11]: access modifier

File: src/main/java/core/basesyntax/figure/Rectangle.java [hunk 23]
```diff
@@ -0,0 +1,55 @@
+package core.basesyntax.figure;
+
+import core.basesyntax.tools.Color;
+
+public class Rectangle extends Figure {
+    private int lengthRectangle;
+    private int widthRectangle;
+    private double area;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Rectangle.java:8]:
```suggestion
    private int length;
    private int width;
```

File: src/main/java/core/basesyntax/figure/IsoscelesTrapezoid.java [hunk 24]
```diff
@@ -0,0 +1,60 @@
+package core.basesyntax.figure;
+
+import core.basesyntax.tools.Color;
+
+public class IsoscelesTrapezoid extends Figure {
+    private int baseSide;
+    private int topSide;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/IsoscelesTrapezoid.java:7]: what is topSide? 2 sides - left and right? if yes, name it just `side`

File: src/main/java/core/basesyntax/tools/ColorSupplier.java [hunk 25]
```diff
@@ -0,0 +1,11 @@
+package core.basesyntax.tools;
+
+import java.util.Random;
+
+public class ColorSupplier {
+
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/tools/ColorSupplier.java:6]: redundant empty line

File: src/main/java/core/basesyntax/Main.java [hunk 26]
```diff
@@ -0,0 +1,21 @@
+package core.basesyntax;
+
+import core.basesyntax.figure.Figure;
+import core.basesyntax.tools.FigureSupplier;
+
+public class Main {
+    private static FigureSupplier figureCreater = new FigureSupplier();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Main.java:7]:
```suggestion
    private static FigureCreator figureCreator = new FigureSupplier();
```

it is good practice to use interface (a more abstract type) as variable type

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Main.java (hunk 13 above)
Comment (nacenik) [reply src/main/java/core/basesyntax/Main.java:12]: not fixed

File: src/main/java/core/basesyntax/figure/Circle.java (hunk 18 above)
Comment (nacenik) [reply src/main/java/core/basesyntax/figure/Circle.java:7]: not fixed

File: src/main/java/core/basesyntax/figure/Circle.java (hunk 19 above)
Comment (nacenik) [reply src/main/java/core/basesyntax/figure/Circle.java:12]: not fixed

File: src/main/java/core/basesyntax/figure/Circle.java (hunk 21 above)
Comment (nacenik) [reply src/main/java/core/basesyntax/figure/Circle.java:50]: not fixed

File: src/main/java/core/basesyntax/figure/Circle.java [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+    private int radius;
+
+    public Circle(String name, Color color, int radius) {
+        super(name, color);
+        this.radius = radius;
+    }
+
+    @Override
+    public double resultArea() {
+        return (2 * Math.PI * radius * radius);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Circle.java:15]:
```suggestion
        return Math.PI * radius * radius;
```

File: src/main/java/core/basesyntax/figure/Figure.java [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+        return colorFigure;
+    }
+
+    public void setColor(Color colorFigure) {
+        this.colorFigure = colorFigure;
+    }
+
+    @Override
+    public void draw() {
+        System.out.print("figure: " + getName());
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Figure.java:34]:
```suggestion
        System.out.print("figure: " + name);

```

File: src/main/java/core/basesyntax/figure/Rectangle.java [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+
+    public Rectangle(String name, Color color, int length, int width) {
+        super(name, color);
+        this.length = length;
+        this.width = width;
+    }
+
+    @Override
+    public double resultArea() {
+        return (this.length * this.width);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Rectangle.java:17]:
```suggestion
        return length * width;

```

File: src/main/java/core/basesyntax/figure/Rectangle.java [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    @Override
+    public void draw() {
+        super.draw();
+        System.out.println(" area: " + resultArea() + " sq.units"
+                + ", length: " + this.length + " units, "
+                + "width: " + this.width + " units, "
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Rectangle.java:25]:
```suggestion
                + ", length: " + length + " units, "
                + "width: " + width + " units, "

```

File: src/main/java/core/basesyntax/figure/IsoscelesTrapezoid.java [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+                * Math.sqrt((this.rightSide * this.rightSide)
+                - ((this.leftSide * this.leftSide) / 4)));
+    }
+
+    @Override
+    public void draw() {
+        super.draw();
+        System.out.println(" area: " + resultArea() + " sq.units"
+                + ", base: " + this.leftSide + " units, "
+                + "side: " + this.rightSide + " units, "
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/IsoscelesTrapezoid.java:28]: same

File: src/main/java/core/basesyntax/figure/IsoscelesTrapezoid.java [hunk 32]
```diff
@@ -0,0 +1,31 @@
...
+        this.leftSide = leftSide;
+        this.rightSide = rightSide;
+    }
+
+    @Override
+    public double resultArea() {
+        return ((1 / 2)
+                * this.leftSide
+                * Math.sqrt((this.rightSide * this.rightSide)
+                - ((this.leftSide * this.leftSide) / 4)));
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/IsoscelesTrapezoid.java:20]: same
remove `this`

File: src/main/java/core/basesyntax/figure/RightTriangle.java [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+    public double resultArea() {
+        return ((this.firstLeg * this.secondLeg) / 2);
+    }
+
+    @Override
+    public void draw() {
+        super.draw();
+        System.out.println(" area: " + resultArea() + " sq.units"
+                + ", firstLeg: " + this.firstLeg + " units, "
+                + "secondLeg: " + this.secondLeg + " units, "
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/RightTriangle.java:25]: same

File: src/main/java/core/basesyntax/figure/RightTriangle.java [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+
+    public RightTriangle(String name, Color color, int firstLeg, int secondLeg) {
+        super(name, color);
+        this.firstLeg = firstLeg;
+        this.secondLeg = secondLeg;
+    }
+
+    @Override
+    public double resultArea() {
+        return ((this.firstLeg * this.secondLeg) / 2);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/RightTriangle.java:17]:
```suggestion
        return firstLeg * secondLeg / 2;

```

File: src/main/java/core/basesyntax/figure/Square.java [hunk 35]
```diff
@@ -0,0 +1,25 @@
...
+    private int side;
+
+    public Square(String name, Color color, int side) {
+        super(name, color);
+        this.side = side;
+    }
+
+    @Override
+    public double resultArea() {
+        return (side * side);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Square.java:15]:
```suggestion
        return side * side;
```

File: src/main/java/core/basesyntax/figure/Square.java [hunk 36]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    public void draw() {
+        super.draw();
+        System.out.println(" area: " + resultArea() + " sq.units"
+                + ", side: " + this.side + " units, "
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/figure/Square.java:22]:
```suggestion
                + ", side: " + side + " units, "

```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**